pages/en/lb4/Dependency-injection.md

```diff
@@ -140,28 +140,101 @@ class InfoController {
 }
 ```
 
+## Optional dependencies
+
+Sometimes the dependencies are optional. For example, the logging level for a Logger provider can have a default value if it is not set (bound to the context). 
+
+To resolve an optional dependency, set `optional` flag to true:
+
+```ts
+const ctx = new Context();
+await ctx.get('optional-key', {optional: true}); // Return `undefined` instead of throwing an error
+```
+
+Here is anthoer example with dependency injection:
+
+```ts
+export class LoggerProvider implements Provider<Logger> {
+  // Log writer is an optional dependency and it falls back to `logToConsole`
+  @inject('log.writer', {optional: true})
+  private logWriter: LogWriterFn = logToConsole;
+
+  @inject('log.level', {optional: true})
+  private logLevel: string = 'WARN';
+}
+```
+
+Optional dependencies can be used with constructor, property, and method injections. You can use default values with such cases. For example:
+
+```ts
+export class LoggerProvider implements Provider<Logger> {
+  constructor(
+    // Log writer is an optional dependency and it falls back to `logToConsole`
+    @inject('log.writer', {optional: true})
+    private logWriter: LogWriterFn = logToConsole,
+
+    @inject('log.level', {optional: true})
+    private logLevel: string = 'WARN',
+  ) {}
+}
+```
+
 ## Circular dependencies
 
 LoopBack can detect circular dependencies and report the path which leads to the problem.
 For example,
 
 ```ts
-const context = new Context();
-  interface XInterface {}
-  interface YInterface {}
+import {Context, inject} from '@loopback/context';
 
-  class XClass implements XInterface {
-    @inject('y') public y: YInterface;
-  }
+interface Developer {
+  // Each developer belongs to a team
+  team: Team;
+}
+
+interface Team {
+  // Each team works on a project
+  project: Project;
+}
+
+interface Project {
+  // Each project has a lead developer
+  lead: Developer;
+}
 
-  class YClass implements YInterface {
-    @inject('x') public x: XInterface;
+class DeveloperImpl implements Developer {
+  constructor(@inject('team') public team: Team) {}
+}
+
+class TeamImpl implements Team {
+  constructor(@inject('project') public project: Project) {}
+}
+
+class ProjectImpl implements Project {
+  constructor(@inject('lead') public lead: Developer) {}
+}
+
+export function main() {
```

Comment:
So far, none of code examples are exporting a `main` function. To be honest, I don't understand why do you need it here anyways, since the body of `main` is synchronous code that can be placed directly in the main scope of the `.js`/`.ts` file.

I am proposing to reorganize this code to avoid `export function main`.

Please let me know if there are good reasons for keeping it, it's entirely possibly I am missing something here.

Reply:
See https://github.com/raymondfeng/loopback4-example-di/blob/master/index.js. My demo app allows the user to choose which one to run from the command line.